### ISSUE TITLE
Create an abstract class in Knora-base ontology

### ISSUE THREAD

**Issue by mrivoal** [dasch-swiss/dsp-api#354]
Allow the creation of abstract classes in  knora-base ontology, for modelling purposes, so that such an abstract class is not displayed as a regular class in the Salsah GUI.

Abstract classes will be used to deal with use cases where we want to use the same property for two different resources while still complying with `knora-base:objectClassConstraint` and `knora-base:subjectClassConstraint`. 

For example a `Commentable` abstract resource class is created, for resources that can have a comment.   `Commentable` has property `has comment` . The property `has comment` has `knora-base:subjectClassConstraint` on `Commentable`. 
We also create `Article` and `Book` classes, both being subclasses of `Commentable`. Thus, `Article` and `Book` classes inherit `has comment` property from their mother-class.

But `Commentable` is a class that is not meant to be seen by GUI users. At least, it is not as a resource a user should see in the window "add resource".

**Comment by benjamingeer** [comment]
I think both classes and properties should have an optional `isAbstract` annotation property, which would tell Knora not to instantiate them directly.

I think we could also write GraphDB consistency rules to check this.

Currently some of this is hard-coded. There's a list of abstract resource classes in `OntologyConstants.KnoraBase.AbstractResourceClasses`, and the ontology responder also has some code that knows about abstract classes and properties.

**Comment by tobiasschweizer** [comment]
So this would be like an abstract class or an interface in Java which cannot be instantiated itself (only by subclassing it)?

**Comment by benjamingeer** [comment]
Yes, that's the idea.

**Comment by tobiasschweizer** [comment]
I think it is a good idea to put that information in the ontology. Then the scala code wouldn't have to be adapted once a new abstract resource class is added. We could simply ignore these resources (evaluating `isAbstract`) in the sparql queries used by ontology responder.

**Comment by benjamingeer** [comment]
Yes, and in the Scala code, we could also check query results to see if we got an abstract class or property (by asking the ontology responder), and if so, we could throw a nice exception.

**Comment by tobiasschweizer** [comment]
So in the end when some tries to instantiate an abstract resource class, the ontology responder would not find anything about it?

E.g., someone wants to create a resource of an abstract type, and the resources responder would ask the ontology responder about the cardinalities of this class. Then the ontology responder would not find this class because `isAbstract` is true and thus it was ignored.

**Comment by benjamingeer** [comment]
I think the ontology responder should know about all the abstract classes and properties. When the resources responder asks for information about a resource class, the ontology responder could check whether it's an abstract class, and if so, it could throw a special exception like `AbstractClassException`, which would be propagated back to the resources responder. That would be better than `NotFoundException`.

**Comment by tobiasschweizer** [comment]
Right, that seems fine to me.

More generally, could we aim at making all knora-base resource classes abstract? To  my understanding, entities defined in knora-base are very generic so they do not "mean" anything.

Still we can make ontologies defining basic concepts that can be reused among different projects (like for the bibliography).

I am thinking about creating a base set of standoff tags in a special ontology and move them out of knora-base. Only the the very basic concepts of standoff would remain there (StandoffTagV1 and the data type standoff tags).

**Comment by benjamingeer** [comment]
> More generally, could we aim at making all knora-base resource classes abstract? To my understanding, entities defined in knora-base are very generic so they do not "mean" anything.

We have to decide which ones should be abstract: #138.

> Still we can make ontologies defining basic concepts that can be reused among different projects (like for the bibliography).

Yes, after they go through some kind of standardisation process.

> I am thinking about creating a base set of standoff tags (like for the TEI) in a special ontology and move them out of knora-base. Only the the very basic concepts of standoff would remain in knora-base (StandoffTagV1 and the data type standoff tags).

Sounds OK to me.

**Comment by benjamingeer** [comment]
I think we might also be able to make GraphDB consistency rules like this:

```
Consistency: direct_use_of_abstract_class
    i <rdf:type> t [Context <onto:explicit>]
    t <knora-base:isAbstract> true
    ------------------------------------

Consistency: direct_use_of_abstract_property
    i p j [Context <onto:explicit>]
    p <knora-base:isAbstract> true
    ------------------------------------
```

**Comment by tobiasschweizer** [comment]
Ok, I moved all non system standoff information to a dedicated ontology: https://github.com/dhlab-basel/Knora/pull/319/commits/35a603f288dbe00acf6678b4076bf62088169758.

I first made a few mistakes when copying the ontology information to another file because I forgot to change some of the `:` prefixes to `knora-base`. 
We need to find a way to check an ontology for unresolvable references. Moreover, I could instantiate standoff tags that were unknown to the application because I did not load `standoff-onto.ttl` in the tests yet.
